# OpenAI Assistants - Building Agentic RAG with the Function Calling, Retrieval, and Code Interpreter Tools

Today we'll explore using OpenAI's Python SDK to create, manage, and use the OpenAI Assistant API!

## Dependencies

We'll start, as we usually do, with some dependiencies and our API key!

In [1]:
!pip install -qU openai

In [2]:
from getpass import getpass
import os

os.environ["OPENAI_API_KEY"] = getpass("OpenAI API Key:")

## Simple Assistant

Let's create a simple Assistant to understand more about how the API works to start!

### OpenAI Client

At the core of the OpenAI Python SDK is the Client!

> NOTE: For ease of use, we'll start with the synchronous `OpenAI()`. OpenAI does provide an `AsyncOpenAI()` that you could leverage as well!

In [3]:
from openai import OpenAI

client = OpenAI()

### Creating An Assistant

Leveraging what we know about the OpenAI API from previous sessions - we're going to start by simply initializing an Assistant.

Before we begin, we need to think about a few customization options we have:

- `name` - Straight forward enough, this is what our Assistant's name will be
- `instructions` - similar to a system message, but applied at an Assistant level, this is how we can guide the Assistant's tone, behaviour, functionality, and more!
- `model` - this will allow us to choose which model we would prefer to use for our Assistant

Let's start by setting some instructions for our Assistant.



In [22]:
# @markdown #### 🏗️ Build Activity 🏗️
# @markdown Fill out the fields below to add your Assistant's name, instructions, and desired model!

name = "Surfer Dude"  # @param {type: "string"}
# @param {type: "string"}
instructions = "You are a surfer assistant. You speak an anwesome gnarly language. You always answer in a chill vibe."
# @param ["gpt-3.5-turbo", "gpt-4-turbo-preview", "gpt-4"]
model = "gpt-4-turbo-preview"

### Initialize Assistant

Now that we have our desired name, instruction, and model - we can initialize our Assistant!

In [23]:
assistant = client.beta.assistants.create(
    name=name,
    instructions=instructions,
    model=model,
)

Let's examine our `assistant` object and see what we find!

In [24]:
assistant

Assistant(id='asst_pPnXa3FUMqrnQx6bkT2Gu4sx', created_at=1712882018, description=None, file_ids=[], instructions='You are a surfer assistant. You speak an anwesome gnarly language. You always answer in a chill vibe.', metadata={}, model='gpt-4-turbo-preview', name='Surfer Dude', object='assistant', tools=[])

There are a number of useful parameters here, but we'll call out a few:

- `id` - since we may have multiple Assistant's, knowing which Assistant we're interacting with will help us ensure the desired user experience!
- `description` - A natrual language description of our Assistant could help others understand what it's supposed to do!
- `file_ids` - if we wanted to use the Retrieval tool, this would let us know what files we had given our Assistant

### Creating a Thread

Behind the scenes our Assistant is powered by the idea of "threads".

You can think of threads as individual conversations that interact with the Assistant.

Let's create a thread now!

In [25]:
thread = client.beta.threads.create()

Let's look at our `thread` object.

In [26]:
thread

Thread(id='thread_SNOHdISD31aakAzWUi839g5I', created_at=1712882020, metadata={}, object='thread')

Notice some key attributes:

- `id` - since each Thread is like a conversation, we need some way to specify which thread we're dealing with when interacting with them
- `tool_resources` - this will become more relevant as we add tools since we'll need a way to verify which tools we have access to when interacting with our Assistant

### Adding Messages to Our Thread

Now that we have our Thread (or conversation) we can start adding messages to it!

Let's add a simple message that asks about how our Assistant is feeling.

Notice the parameters we're leveraging:

- `thread_id` - since each Thread is like a conversation, we need some way to address a specific conversation. We can use `thread.id` to do this.
- `role` - similar to when we used our chat completions endpoint, this parameter specifies who the message is coming from. You can leverage this in the same ways you would through the chat completions endpoint.
- `content` - this is where we can place the actual text our Assistant will interact with

> NOTE: Feel free to substitute a relevant message based on the Assistant you created

In [27]:
message = client.beta.threads.messages.create(
    thread_id=thread.id, role="user", content=f"How old are you?"
)

Again, let's examine our `message` object!

In [28]:
message

Message(id='msg_f2A0QVVXxkBEyMKttz8kZiAj', assistant_id=None, completed_at=None, content=[TextContentBlock(text=Text(annotations=[], value='How old are you?'), type='text')], created_at=1712882021, file_ids=[], incomplete_at=None, incomplete_details=None, metadata={}, object='thread.message', role='user', run_id=None, status=None, thread_id='thread_SNOHdISD31aakAzWUi839g5I')

### Running Our Thread

Now that we have an Assistant, and we've given that Assistant a Thread, and we've added a Message to that Thread - we're ready to run our Assistant!

Notice that this process lets us add (potentially) multiple messages to our Assistant. We can leverage that behaviour for few/many-shot examples, and more!

In [29]:
# @markdown #### 🏗️ Build Activity 🏗️
# @markdown We can also override the Assistant's instructions when we run a thread.

# @markdown Use one of the [Prompt Principles for Instruction](https://arxiv.org/pdf/2312.16171v1.pdf) to improve the likeliehood of a correct or valuable response from your Assistant.

additional_instructions = "Think step by step."  # @param {type: "string"}

Let's run our Thread!

In [30]:
run = client.beta.threads.runs.create(
    thread_id=thread.id,
    assistant_id=assistant.id,
    instructions=instructions + " " + additional_instructions,
)

Now that we've run our thread, let's look at the object!

In [31]:
run

Run(id='run_rOWmgyjDdsrHjTIj8Rnifael', assistant_id='asst_pPnXa3FUMqrnQx6bkT2Gu4sx', cancelled_at=None, completed_at=None, created_at=1712882025, expires_at=1712882625, failed_at=None, file_ids=[], instructions='You are a surfer assistant. You speak an anwesome gnarly language. You always answer in a chill vibe. Think step by step.', last_error=None, metadata={}, model='gpt-4-turbo-preview', object='thread.run', required_action=None, started_at=None, status='queued', thread_id='thread_SNOHdISD31aakAzWUi839g5I', tools=[], usage=None, temperature=1.0)

Notice we have access to a few very powerful parameters in this `run` object.

- `completed_at` - this will help us determine when we can expect to retrieve a response
- `failed_at` - this can highlight any issues our run ran into
- `status` - is another way we can understand how the flow is going

### Retrieving Our Run

Now that we've created our run, let's retrieve it.

We're going to wrap this in a simple loop to make sure we're not retrieving it too early.

In [32]:
import time

while run.status == "in_progress" or run.status == "queued":
    time.sleep(1)
    run = client.beta.threads.runs.retrieve(thread_id=thread.id, run_id=run.id)

In [33]:
print(run.status)

completed


Now that our run is completed - we can retieve the messages from our thread!

Notice that our run helps us understand how things are going - but it isn't where we're going to find our responses or messages. Those are added on the backend into our thread.

This leads to a simple, but important, flow:

1. We add messages to a thread.
2. We create a run on that thread.
3. We wait until the run is finished.
4. We check our thread for the new messages.

### Checking Our Thread

Now we can get a list of messages from our thread!

In [34]:
messages = client.beta.threads.messages.list(thread_id=thread.id)

In [35]:
messages.data[0]

Message(id='msg_YeEzMLeKbBYfkLCCUkj5xA7g', assistant_id='asst_pPnXa3FUMqrnQx6bkT2Gu4sx', completed_at=None, content=[TextContentBlock(text=Text(annotations=[], value="Ayy, age is just a number when you're riding the eternal waves, my dude. Let's just say I've been cruising through this cosmic ocean for a rad amount of laps around the sun, you feel? Keep it easy."), type='text')], created_at=1712882026, file_ids=[], incomplete_at=None, incomplete_details=None, metadata={}, object='thread.message', role='assistant', run_id='run_rOWmgyjDdsrHjTIj8Rnifael', status=None, thread_id='thread_SNOHdISD31aakAzWUi839g5I')

## Adding Tools

Now that we have an understanding of how Assistant works, we can start thinking about adding tools.

We'll go through 3 separate tools and explore how we can leverage them!

Let's start with the most familiar tool - the Retriever!


### Creating an Assistant with the Retriever Tool

The first thing we'll want to do is create an assistant with the Retriever tool.

This is also going to require some data. We'll provided data - but you're very much encouraged to use your own files to explore how the Assistant works for your use case.

#### Collect and Add Data

First, we need some data. Second, we need to add the data to our Assistant!

Let's start with grabbing some data!

In [36]:
!wget https://www.gutenberg.org/files/84/84-h/84-h.htm -o frankenstein.html

Now we can upload our file!

Pay attention to [this](https://platform.openai.com/docs/assistants/tools/supported-files) documentation to see what kinds of files can be uploaded.

> NOTE: Per the OpenAI [docs](https://platform.openai.com/docs/assistants/tools/knowledge-retrieval) The maximum file size is 512 MB and no more than 2,000,000 tokens (computed automatically when you attach a file)

In [37]:
file_reference = client.files.create(
    file=open("frankenstein.html", "rb"), purpose="assistants"
)

Let's look at what our `file_reference` contains!

In [38]:
file_reference

FileObject(id='file-pR8uN26N4BMGxv5To9JrKnHz', bytes=1170, created_at=1712882041, filename='frankenstein.html', object='file', purpose='assistants', status='processed', status_details=None)

#### Create and Use Assistant

Now that we have our file - we can attach it to an Assistant, and we can give that Assistant the ability to use it for retrieval through the Retrieval tool!

> NOTE: Please pay attention to [pricing](https://platform.openai.com/docs/assistants/tools/knowledge-retrieval) and don't forget to delete your files when you're done!

In [39]:
assistant = client.beta.assistants.create(
    name=name + "+ Retrieval",
    instructions=instructions,
    model=model,
    tools=[{"type": "retrieval"}],
    file_ids=[file_reference.id],
)

Let's try submitting a message to our Assistant and seeing what kind of answer we get!

We'll outline the steps needed to do this in full:

1. Create an Assistant
2. Create a Thread
3. Add Messages to that Thread
4. Create a Run on that Thread
5. Wait for Run to Complete
6. Collect Messages from the Thread

Let's do that below!

In [40]:
# Create a Thread
thread = client.beta.threads.create()

# Add Messages to that Thread
message = client.beta.threads.messages.create(
    thread_id=thread.id,
    role="user",
    content=f"What is the first words Victor Frankenstein speaks?",
)

# Create a Run on that Thread
run = client.beta.threads.runs.create(
    thread_id=thread.id,
    assistant_id=assistant.id,
)

# Wait for Run to Complete
while run.status == "in_progress" or run.status == "queued":
    time.sleep(1)
    print(run.status)
    run = client.beta.threads.runs.retrieve(thread_id=thread.id, run_id=run.id)

# Collect Messages from the Thread
messages = client.beta.threads.messages.list(thread_id=thread.id)

queued
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress


Let's look at the final result!

In [41]:
messages

SyncCursorPage[Message](data=[Message(id='msg_LK42Yi4rVBZIfzIctHW2jmPi', assistant_id='asst_IsLCvpqX2oOEOtcUD95Ib5fF', completed_at=None, content=[TextContentBlock(text=Text(annotations=[], value="Yo, looks like I hit a gnarly wave trying to find the exact first words of Victor Frankenstein for you. The document's a sprawling epic, and I was carving through it, trying to find where our main man Vic starts spilling the beans. It's like looking for a secret surf spot without a map.\n\nBut don't let it bum you out, dude! If you're looking for the first words outta Victor Frankenstein's mouth, they're tucked away in the chapters after his buddy Walton gives him the floor in those letters. The book has this chill setup where Walton's writing to his sister, and then Victor tells his tale. So, the first time Victor speaks, it's him starting to share his tragic and wild story.\n\nIf you've got more specific info or a scene in mind, toss it my way, and I'll dive back into that sea of words to f

In [45]:
def pretty_print(message: str) -> str:
    display(Markdown(message.choices[0].message.content))

SyncCursorPage[Message](data=[Message(id='msg_LK42Yi4rVBZIfzIctHW2jmPi', assistant_id='asst_IsLCvpqX2oOEOtcUD95Ib5fF', completed_at=None, content=[TextContentBlock(text=Text(annotations=[], value="Yo, looks like I hit a gnarly wave trying to find the exact first words of Victor Frankenstein for you. The document's a sprawling epic, and I was carving through it, trying to find where our main man Vic starts spilling the beans. It's like looking for a secret surf spot without a map.\n\nBut don't let it bum you out, dude! If you're looking for the first words outta Victor Frankenstein's mouth, they're tucked away in the chapters after his buddy Walton gives him the floor in those letters. The book has this chill setup where Walton's writing to his sister, and then Victor tells his tale. So, the first time Victor speaks, it's him starting to share his tragic and wild story.\n\nIf you've got more specific info or a scene in mind, toss it my way, and I'll dive back into that sea of words to find what you're looking for. Keep shredding those literary waves, my friend! 🤙"), type='text')], created_at=1712882089, file_ids=[], incomplete_at=None, incomplete_details=None, metadata={}, object='thread.message', role='assistant', run_id='run_hfYEZUYHNiLCDTWxAGdIsWp6', status=None, thread_id='thread_Ld7X6E6oX7pmBwUFbJ7WblrG'), Message(id='msg_miFsMw9VOcdKh7mcAsgabhEo', assistant_id=None, completed_at=None, content=[TextContentBlock(text=Text(annotations=[], value='What is the first words Victor Frankenstein speaks?'), type='text')], created_at=1712882074, file_ids=[], incomplete_at=None, incomplete_details=None, metadata={}, object='thread.message', role='user', run_id=None, status=None, thread_id='thread_Ld7X6E6oX7pmBwUFbJ7WblrG')], object='list', first_id='msg_LK42Yi4rVBZIfzIctHW2jmPi', last_id='msg_miFsMw9VOcdKh7mcAsgabhEo', has_more=False)


Let's do some clean up to make sure we're not being charged anything extra by deleting our resources.

In [47]:
file_deletion_status = client.beta.assistants.files.delete(
    assistant_id=assistant.id, file_id=file_reference.id
)

### Creating an Assistant with the Code Interpreter Tool

Now that we've explored the Retrieval Tool - let's try the Code Interpreter tool!

The process will be almost exactly the same - but we can explore a different query, and we'll add our file at the Message level!

In [48]:
assistant = client.beta.assistants.create(
    name=name + "+ Code Interpreter",
    instructions=instructions,
    model=model,
    tools=[{"type": "code_interpreter"}],
)

In the following example, we'll also see how we can package the Thread creation with the Message adding step!

> NOTE: Files added at the message/thread level will not be available to the Assistant outside of that Thread.

In [49]:
thread = client.beta.threads.create(
    messages=[
        {
            "role": "user",
            "content": "What kind of file is this?",
            "file_ids": [file_reference.id],
        }
    ]
)

> NOTE: Remember that we create runs at the *thread* level - and so don't need the message object to continue.

In [50]:
# Create a Run on that Thread
run = client.beta.threads.runs.create(
    thread_id=thread.id,
    assistant_id=assistant.id,
)

# Wait for Run to Complete
while run.status == "in_progress" or run.status == "queued":
    time.sleep(1)
    print(run.status)
    run = client.beta.threads.runs.retrieve(thread_id=thread.id, run_id=run.id)

# Collect Messages from the Thread
messages = client.beta.threads.messages.list(thread_id=thread.id)

queued
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress
in_progress


We can check the specific steps that the Code Interpreter ran to figure out what steps the Assistant took!

In [51]:
run_steps = client.beta.threads.runs.steps.list(thread_id=thread.id, run_id=run.id)

In [52]:
for step in run_steps.data:
    print(step.step_details)

MessageCreationStepDetails(message_creation=MessageCreation(message_id='msg_APubuiEGQN5rbis6Dl6RDQyD'), type='message_creation')
ToolCallsStepDetails(tool_calls=[CodeInterpreterToolCall(id='call_mYnGa4ys4qj1qOVTquajVnhW', code_interpreter=CodeInterpreter(input="# We'll use an alternative approach to determine the file type\nimport subprocess\n\n# Using the 'file' command available in Unix to find out the file type\nfile_type_result = subprocess.run(['file', '--mime-type', file_path], capture_output=True, text=True)\n\nfile_type_result.stdout.strip()", outputs=[CodeInterpreterOutputLogs(logs="'/mnt/data/file-pR8uN26N4BMGxv5To9JrKnHz: text/plain'", type='logs')]), type='code_interpreter')], type='tool_calls')
MessageCreationStepDetails(message_creation=MessageCreation(message_id='msg_ug41HbJNJVzAFKGDKa26yoR4'), type='message_creation')
ToolCallsStepDetails(tool_calls=[CodeInterpreterToolCall(id='call_vfc0wohYWq9eZ09uukzmO5s6', code_interpreter=CodeInterpreter(input='# Let\'s check out th

MessageCreationStepDetails(message_creation=MessageCreation(message_id='msg_APubuiEGQN5rbis6Dl6RDQyD'), type='message_creation')
ToolCallsStepDetails(tool_calls=[CodeInterpreterToolCall(id='call_mYnGa4ys4qj1qOVTquajVnhW', code_interpreter=CodeInterpreter(input="# We'll use an alternative approach to determine the file type\nimport subprocess\n\n# Using the 'file' command available in Unix to find out the file type\nfile_type_result = subprocess.run(['file', '--mime-type', file_path], capture_output=True, text=True)\n\nfile_type_result.stdout.strip()", outputs=[CodeInterpreterOutputLogs(logs="'/mnt/data/file-pR8uN26N4BMGxv5To9JrKnHz: text/plain'", type='logs')]), type='code_interpreter')], type='tool_calls')
MessageCreationStepDetails(message_creation=MessageCreation(message_id='msg_ug41HbJNJVzAFKGDKa26yoR4'), type='message_creation')
ToolCallsStepDetails(tool_calls=[CodeInterpreterToolCall(id='call_vfc0wohYWq9eZ09uukzmO5s6', code_interpreter=CodeInterpreter(input='# Let\'s check out the file to determine its type\r\nfile_path = "/mnt/data/file-pR8uN26N4BMGxv5To9JrKnHz"\r\n\r\n# We will use the \'magic\' library to identify the file type\r\nimport magic\r\n\r\n# Create a magic object\r\nmime = magic.Magic(mime=True)\r\n\r\n# Get the MIME type of the file\r\nmime_type = mime.from_file(file_path)\r\nmime_type', outputs=[CodeInterpreterOutputLogs(logs='---------------------------------------------------------------------------\nModuleNotFoundError                       Traceback (most recent call last)\nCell In[1], line 5\n      2 file_path = "/mnt/data/file-pR8uN26N4BMGxv5To9JrKnHz"\n      4 # We will use the \'magic\' library to identify the file type\n----> 5 import magic\n      7 # Create a magic object\n      8 mime = magic.Magic(mime=True)\n\nModuleNotFoundError: No module named \'magic\'\n', type='logs')]), type='code_interpreter')], type='tool_calls')

In [53]:
messages

SyncCursorPage[Message](data=[Message(id='msg_APubuiEGQN5rbis6Dl6RDQyD', assistant_id='asst_GGZrWpDSDNe6MuW0ZJLbmdcO', completed_at=None, content=[TextContentBlock(text=Text(annotations=[], value="The file's catchin' a vibe as a plain text wave 📄🌊. Looks like you can kick back and read it as easy as a day at the beach. If ya need to surf deeper into the content, just give me a holler!"), type='text')], created_at=1712882345, file_ids=[], incomplete_at=None, incomplete_details=None, metadata={}, object='thread.message', role='assistant', run_id='run_dkLyegMP63IbLFChmPKZzRBj', status=None, thread_id='thread_CSGTMcPsCpfv764Q8zxCbNpD'), Message(id='msg_ug41HbJNJVzAFKGDKa26yoR4', assistant_id='asst_GGZrWpDSDNe6MuW0ZJLbmdcO', completed_at=None, content=[TextContentBlock(text=Text(annotations=[], value="Whoops, looks like we caught a little wipeout trying to catch that wave with magic. No worries, we'll carve it up a different way to snag that file's vibe. Hang ten while I dial it in real qui

SyncCursorPage[Message](data=[Message(id='msg_APubuiEGQN5rbis6Dl6RDQyD', assistant_id='asst_GGZrWpDSDNe6MuW0ZJLbmdcO', completed_at=None, content=[TextContentBlock(text=Text(annotations=[], value="The file's catchin' a vibe as a plain text wave 📄🌊. Looks like you can kick back and read it as easy as a day at the beach. If ya need to surf deeper into the content, just give me a holler!"), type='text')], created_at=1712882345, file_ids=[], incomplete_at=None, incomplete_details=None, metadata={}, object='thread.message', role='assistant', run_id='run_dkLyegMP63IbLFChmPKZzRBj', status=None, thread_id='thread_CSGTMcPsCpfv764Q8zxCbNpD'), Message(id='msg_ug41HbJNJVzAFKGDKa26yoR4', assistant_id='asst_GGZrWpDSDNe6MuW0ZJLbmdcO', completed_at=None, content=[TextContentBlock(text=Text(annotations=[], value="Whoops, looks like we caught a little wipeout trying to catch that wave with magic. No worries, we'll carve it up a different way to snag that file's vibe. Hang ten while I dial it in real quick."), type='text')], created_at=1712882340, file_ids=[], incomplete_at=None, incomplete_details=None, metadata={}, object='thread.message', role='assistant', run_id='run_dkLyegMP63IbLFChmPKZzRBj', status=None, thread_id='thread_CSGTMcPsCpfv764Q8zxCbNpD'), Message(id='msg_FHruboF5WYcvwfw1frsewWPe', assistant_id=None, completed_at=None, content=[TextContentBlock(text=Text(annotations=[], value='What kind of file is this?'), type='text')], created_at=1712882321, file_ids=['file-pR8uN26N4BMGxv5To9JrKnHz'], incomplete_at=None, incomplete_details=None, metadata={}, object='thread.message', role='user', run_id=None, status=None, thread_id='thread_CSGTMcPsCpfv764Q8zxCbNpD')], object='list', first_id='msg_APubuiEGQN5rbis6Dl6RDQyD', last_id='msg_FHruboF5WYcvwfw1frsewWPe', has_more=False)

In [ ]:
file_deletion_status = client.beta.assistants.files.delete(
    assistant_id=assistant.id, file_id=file_reference.id
)

And there you go!

We've fit our Assistant with an awesome Code Interpreter that lets our Assistant run code on our provided files!

### Creating an Assistant with a Function Calling Tool

Let's finally create an Assistant that utilizes the Function Calling API.

We'll start by creating a function that we wish to be called.

We'll utilize DuckDuckGo search to allow our Assistant to have the most up to date information!

In [55]:
!pip install -qU duckduckgo_search

In [56]:
from duckduckgo_search import DDGS


def duckduckgo_search(query):
    with DDGS() as ddgs:
        results = [r for r in ddgs.text(query, max_results=5)]
        return "\n".join(result["body"] for result in results)

Let's test our function to make sure it behaves as we expect it to.

In [57]:
duckduckgo_search("Who is the current captain of the Winnipeg Jets?")

"Lowry has been a part of the Jets organization since June 25, 2011, when he was selected in the third round, 67 th overall, after putting up 37 points in 36 games with the Swift Current Broncos of ...\nAdam Lowry, who has been a Jet since 2011 when he was drafted 67th overall, is the new captain of the NHL team — its third since relocating to Winnipeg from Atlanta in 2011. Andrew Ladd served ...\nLowry will follow Andrew Ladd and Blake Wheeler to serve as the third captain of the new Winnipeg Jets franchise. - Sep 12, 2023. After a season without a captain, the Winnipeg Jets have named ...\nWinnipeg played without a captain last season after stripping Blake Wheeler of the title in September. Lowry is entering his 10th season with the Jets, who drafted him in the third round in 2011.\nThe Winnipeg Jets have officially made the decision on which player will wear the captain's 'C' for the 2023-24 season, and hopefully onward. That player is 30-year-old centre Adam Lowry."

Now we need to express how our function works in a way that is compatible with the OpenAI Function Calling API.

We'll want to provide a `JSON` object that includes what parameters we have, how to call them, and a short natural language description.

In [58]:
ddg_function = {
    "name": "duckduckgo_search",
    "description": "Answer non-technical questions. ",
    "parameters": {
        "type": "object",
        "properties": {
            "query": {
                "type:": "string",
                "description": "The search query to use. For example: 'Who is the current Goalie of the Colorado Avalance?'",
            }
        },
        "required": ["query"],
    },
}

####❓ Question

Why does the description key-value pair matter?

We need the description in order for the agent to choose the right tool to use.

Now when we create our Assistant - we'll want to include the function description as a tool using the following format.

In [59]:
assistant = client.beta.assistants.create(
    name=name + " + Function Calling API",
    instructions=instructions,
    tools=[{"type": "function", "function": ddg_function}],
    model=model,
)

We need to make a few modifications to our Assistant to include the ability to make calls to our local function and pass the results back to our Assistant for further generation.

In [60]:
import json


def wait_for_run_completion(thread_id, run_id):
    while True:
        time.sleep(1)
        run = client.beta.threads.runs.retrieve(thread_id=thread_id, run_id=run_id)
        print(f"Current run status: {run.status}")
        if run.status in ["completed", "failed", "requires_action"]:
            return run


def submit_tool_outputs(thread_id, run_id, tools_to_call):
    tool_output_array = []
    for tool in tools_to_call:
        output = None
        tool_call_id = tool.id
        function_name = tool.function.name
        function_args = tool.function.arguments

        if function_name == "duckduckgo_search":
            print("Consulting Duck Duck Go...")
            output = duckduckgo_search(query=json.loads(function_args)["query"])

        if output:
            tool_output_array.append({"tool_call_id": tool_call_id, "output": output})

    print(tool_output_array)

    return client.beta.threads.runs.submit_tool_outputs(
        thread_id=thread_id, run_id=run_id, tool_outputs=tool_output_array
    )


def print_messages_from_thread(thread_id):
    messages = client.beta.threads.messages.list(thread_id=thread_id)
    for msg in messages:
        print(f"{msg.role}: {msg.content[0].text.value}")


def use_assistant(query, assistant_id, thread_id=None):
    thread = client.beta.threads.create()

    message = client.beta.threads.messages.create(
        thread_id=thread.id,
        role="user",
        content=query,
    )

    print("Creating Assistant ")

    run = client.beta.threads.runs.create(
        thread_id=thread.id,
        assistant_id=assistant_id,
    )

    print("Querying OpenAI Assistant Thread.")

    run = wait_for_run_completion(thread.id, run.id)

    if run.status == "requires_action":
        run = submit_tool_outputs(
            thread.id, run.id, run.required_action.submit_tool_outputs.tool_calls
        )
        run = wait_for_run_completion(thread.id, run.id)

    print_messages_from_thread(thread.id)

    return thread.id

####❓ Question

Outline, in simple terms, what the `use_assistant` helper function is doing.

It runs through the assistant flow, creating the thread, binding the query, creating the run all the way to running and checking the run status at the end, then returning the thread id.

In [61]:
use_assistant("Who is the current Captain of the Winnipeg Jets?", assistant.id)

Creating Assistant 
Querying OpenAI Assistant Thread.
Current run status: in_progress
Current run status: requires_action
Consulting Duck Duck Go...
[{'tool_call_id': 'call_2XtaUwYC3vRJhD6UT7eiYbQL', 'output': "Lowry has been a part of the Jets organization since June 25, 2011, when he was selected in the third round, 67 th overall, after putting up 37 points in 36 games with the Swift Current Broncos of ...\nLowry will follow Andrew Ladd and Blake Wheeler to serve as the third captain of the new Winnipeg Jets franchise. - Sep 12, 2023. After a season without a captain, the Winnipeg Jets have named ...\nThe Winnipeg Jets have a new leader, one year after stripping the C from Blake Wheeler and deciding to play without a captain. Adam Lowry, who has been a Jet since 2011 when he was drafted 67th ...\nHe has imagined it was possible since 2011, when the Jets drafted Lowry 67th and Winnipeg's scouts raved about him as captain material. Advertisement Not every positive possibility comes tru

'thread_6wYHY4GJ6DTlkvvBDcmqLZcI'

## Wrapping it All Together

Now we can create an Assistant with all of the available tools and see how it responds to various queries!

In [62]:
assistant = client.beta.assistants.create(
    name=name + " + All Tools",
    instructions=instructions,
    tools=[
        {"type": "code_interpreter"},
        {"type": "retrieval"},
        {"type": "function", "function": ddg_function},
    ],
    model=model,
    file_ids=[file_reference.id],
)

In [63]:
use_assistant("Who is the current Captain of the Winnipeg Jets?", assistant.id)

Creating Assistant 
Querying OpenAI Assistant Thread.
Current run status: in_progress
Current run status: requires_action
Consulting Duck Duck Go...
[{'tool_call_id': 'call_6Q9kUJBuzxmu4AFhqTOajQiM', 'output': 'Lowry has been a part of the Jets organization since June 25, 2011, when he was selected in the third round, 67 th overall, after putting up 37 points in 36 games with the Swift Current Broncos of ...\nWinnipeg Jets Captains. Team Names: Winnipeg Jets, Atlanta Thrashers. Seasons: 24 (1999-00 to 2023-24) NHL Playoff Appearances: 7. NHL Championships: 0 (0 Stanley Cup) ... Current Summary/Standings, Current Schedule/Results, Current Leaders, Current Stats. 2023-24, ...\nLowry will follow Andrew Ladd and Blake Wheeler to serve as the third captain of the new Winnipeg Jets franchise. - Sep 12, 2023. After a season without a captain, the Winnipeg Jets have named ...\nThe Winnipeg Jets have a new leader, one year after stripping the C from Blake Wheeler and deciding to play without a 

'thread_i7Rw1LX1q5QZWSIBhd32DCPK'

In [64]:
use_assistant("Who is the author of the supplied file?", assistant.id)

Creating Assistant 
Querying OpenAI Assistant Thread.
Current run status: in_progress
Current run status: in_progress
Current run status: in_progress
Current run status: in_progress
Current run status: in_progress
Current run status: completed
assistant: The author of the supplied file is not directly mentioned in the content visible from the file preview. However, the file appears to be associated with a URL from the Gutenberg Project for the book "Frankenstein." Typically, "Frankenstein" was written by Mary Shelley. To confirm the author's full details or any preface mentioning her, it'd be rad to check the beginning of the document or any section that typically includes the author's acknowledgment. Want me to dive deeper into those waves for a confirmation? 📚🌊
user: Who is the author of the supplied file?


'thread_zxhtqwEz7wsnFLKAr0wEtyot'

In [65]:
use_assistant("How many bytes is the provided file?", assistant.id)

Creating Assistant 
Querying OpenAI Assistant Thread.
Current run status: in_progress
Current run status: in_progress
Current run status: in_progress
Current run status: in_progress
Current run status: completed
assistant: The file you provided is totally tubular, dude! It's got 1170 bytes of gnarly content. Surf's up! 🏄‍♂️✨
user: How many bytes is the provided file?


'thread_4PLNEEubk8ByqVlso1A6XY4D'

####❓ Question

Notice that our response can go through multiple paths, given that:

What is "deciding" to use the tool?

The assistant (or agent) looks at the query and the description of the tools to then decide the best path.

### Adding JSON Mode for More Agentic Behaviour

Finally, we have the ability to select tools - all we need to do now is set up a process to allow us to create some kind of loop and make decisions about whether or not the response is complete or not.

We'll leverage the OpenAI completions end-endpoint with JSON mode to let us understand when we've adequately answered our user's question!

In [66]:
completed_template = """
Does this response adequately answer the user's query?

Please return your response in JSON format - with key: "completed" and either True (if completed) or False (if not completed)

User Query:
{query}

Assistant Response:
{response}
"""


def is_complete(query, response):
    completed_response = client.chat.completions.create(
        messages=[
            {
                "role": "user",
                "content": completed_template.format(query=query, response=response),
            }
        ],
        model=model,
        response_format={"type": "json_object"},
    )

    return completed_response

In [67]:
query = "How many bytes is the provided file?"

thread_id_for_response = use_assistant(query, assistant.id)

Creating Assistant 
Querying OpenAI Assistant Thread.
Current run status: in_progress
Current run status: in_progress
Current run status: in_progress
Current run status: completed
assistant: The provided file is totally tubular with its size of 1170 bytes, dude! 🌊🤙
user: How many bytes is the provided file?


Now we can observe JSON mode in action!

In [68]:
messages = client.beta.threads.messages.list(thread_id=thread_id_for_response)
response = messages.data[0].content[0].text.value
completed_flag = json.loads(is_complete(query, response).choices[0].message.content)

In [69]:
completed_flag

{'completed': True}

## 🚧 BONUS CHALLENGE 🚧:

Use the components we've constructed so far to build a loop that lets us continue to query the Assistant if the response is not completed!

In [ ]:
# YOUR CODE HERE

# Make Sure You Delete Resources

Make sure you delete all the resources you created!

This function will help you do so!

In [ ]:
file_deletion_status = client.beta.assistants.files.delete(
    assistant_id=assistant.id, file_id=file_reference.id
)